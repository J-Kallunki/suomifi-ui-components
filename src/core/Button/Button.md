```js
<Button
  className="pizza-button"
  onClick={() => console.log('Gimme pizza!')}
  testId="pizza-test"
>
  🍕
</Button>
```
