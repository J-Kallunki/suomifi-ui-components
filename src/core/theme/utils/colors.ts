import { ColorDesingTokens } from 'suomifi-design-tokens';
import { InternalTokensProp } from '../';

export const colorValue = (props: InternalTokensProp) => (
  colorToken: keyof ColorDesingTokens,
) => props.tokens.colors[colorToken].hsl;

export const colorsUtils = (colors: ColorDesingTokens) =>
  Object.entries(colors).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: value.hsl,
    }),
    {} as { [key in keyof ColorDesingTokens]: string },
  );
