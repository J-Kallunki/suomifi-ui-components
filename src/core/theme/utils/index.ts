export { withDefaultTheme } from './defaultTheme';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding } from './spacing';
