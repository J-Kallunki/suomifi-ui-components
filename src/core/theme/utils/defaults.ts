import {
  suomifiTheme,
  SuomifiTheme,
  defaultTokens,
  SuomifiTokens,
  DefaultSuomifiTokens,
  TokensProp,
} from '../';
import { asPropType } from '../../../utils/typescript';

const internalTokens = (props: TokensProp) =>
  !!props.tokens ? { ...defaultTokens, ...props.tokens } : defaultTokens;

/**
 * Check component props and do common defaulting
 * - Check if tokens are given or use default ones
 * - include internal tokens (THESE ARE NOT GEMERATED BY GIVEN TOKENS! but can be overridden by given tokens)
 * - If as-prop (styled-components) set pass it as asProp
 * @param props All component's props
 */
export const withSuomifiDefaultProps = <
  T extends TokensProp & { asProp?: asPropType }
>({
  tokens,
  as,
  ...props
}: Partial<T> & { as?: asPropType }): T & { tokens: DefaultSuomifiTokens } =>
  ({
    ...props,
    tokens: internalTokens({ tokens }),
    ...(!!as ? { asProp: as } : {}),
  } as T & { tokens: DefaultSuomifiTokens });

export interface Tokens {
  tokens: SuomifiTokens;
}

export interface Theme {
  theme: SuomifiTheme;
}
export type TokensOrThemeProps = Tokens | Theme;

/**
 * Return theme if defined or theme based on tokens
 * @param {Object} prop.theme Theme if defined
 * @param {Object} prop.tokens Tokens
 */
export const themeOrTokens = (props: TokensOrThemeProps) =>
  'theme' in props ? props.theme : suomifiTheme(props.tokens);
