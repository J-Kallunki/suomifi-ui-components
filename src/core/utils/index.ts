import { FlattenSimpleInterpolation } from 'styled-components';
import { suomifiTheme, SuomifiThemeProp } from '../theme';

interface BaseStylesInterface {
  theme: SuomifiThemeProp;
  [key: string]: any;
}

/**
 *  Get CSS export from baseStyles generated by Emotion/Styled
 * @param baseStyles Emotion CSS
 * @param props properties needed to get desired styles from components
 */
export const cssFromBaseStyles = (
  baseStyles: ({ theme }: BaseStylesInterface) => FlattenSimpleInterpolation,
  props = {},
) => baseStyles({ ...props, theme: suomifiTheme() }).join('');
