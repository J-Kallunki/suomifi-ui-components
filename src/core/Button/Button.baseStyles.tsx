import { css } from '@emotion/core';
import { suomifiTheme, ThemeProp } from '../theme';
import { ButtonProps } from './Button';
import { element, focus, fonts } from '../theme/reset';

const fullWidthStyles = css`
  display: block;
  width: 100%;
`;

const negativeStyles = (theme: ThemeProp) => css`
  background: none;
  background-color: ${theme.colors.highlightBase};
  border: 1px solid ${theme.colors.whiteBase};
  text-shadow: none;

  &:hover {
    background: ${theme.gradients.whiteBaseNegative};
  }

  &:active {
    background: none;
    background-color: ${theme.colors.highlightBase};
  }

  &[disabled],
  &:disabled {
    opacity: 0.5;
    background: none;
  }
`;

const secondaryStyles = (theme: ThemeProp) => css`
  color: ${theme.colors.highlightBase};
  background: none;
  background-color: ${theme.colors.whiteBase};
  border: 1px solid ${theme.colors.highlightBase};
  text-shadow: none;

  &:hover {
    background: ${theme.gradients.highlightLight45};
  }

  &:active {
    background: none;
    background-color: ${theme.colors.highlightLight53};
  }

  &[disabled],
  &:disabled {
    color: ${theme.colors.depthBase};
    text-shadow: none;
    background: none;
    background-color: ${theme.colors.highlightLight53};
    border-color: ${theme.colors.depthBase};
  }
`;

const secondaryNoBorderStyles = (theme: ThemeProp) => css`
  ${secondaryStyles(theme)}
  border: none;
`;

const tertiaryStyles = (theme: ThemeProp) => css`
  ${secondaryStyles(theme)}
  background: ${theme.colors.highlightLight50};
  border: none;

  &:hover {
    background: ${theme.colors.highlightLight53};
  }

  &:active {
    background: ${theme.colors.highlightLight53};
  }
`;

export const baseStyles = ({
  theme = suomifiTheme,
  fullWidth = false,
  variant = 'default',
}: ButtonProps) => css`
  ${element}
  ${fonts.inputSemibold}
  ${focus}
  padding: 10px 20px;
  min-height: 40px;
  color: ${theme.colors.whiteBase};
  background: ${theme.gradients.highlightBase};
  border-radius: ${theme.radius.basic};
  text-align: center;
  text-shadow: ${theme.shadows.invertTextShadow};
  cursor: pointer;

  &:hover {
    background: ${theme.gradients.highlightLight4};
  }

  &:active {
    background: ${theme.gradients.highlightDark9};
  }

  &[disabled],
  &:disabled {
    text-shadow: 0 1px 1px ${theme.colors.blackBase};
    background: ${theme.gradients.depthBase};
    pointer-events: none;
    user-select: none;
  }

  ${fullWidth && fullWidthStyles}
  ${variant === 'negative' && negativeStyles(theme)}
  ${variant === 'secondary' && secondaryStyles(theme)}
  ${variant === 'secondary-noborder' && secondaryNoBorderStyles(theme)}
  ${variant === 'tertiary' && tertiaryStyles(theme)}

  & > .fi-button-icon {
    width: 16px;
    height: 16px;
    margin-right: 8px;
    &.fi-button-icon--right {
      margin-right: 0;
      margin-left: 8px;
    }
  }
`;

export const unStyled = ({ theme = suomifiTheme }) => css`
  ${element}
  ${focus}
  border-radius: ${theme.radius.basic};
  cursor: pointer;
`;
