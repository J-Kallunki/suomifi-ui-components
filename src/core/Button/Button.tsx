import React, { Component } from 'react';
import styled from '@emotion/styled';
import classnames from 'classnames';
import { ThemeComponent, ThemeProp } from '../theme';
import { withDefaultTheme } from '../theme/utils/defaultTheme';
import { baseStyles, iconBaseStyles } from './Button.baseStyles';
import {
  Button as CompButton,
  ButtonProps as CompButtonProps,
} from '../../components/Button/Button';
import { Icon, IconProps, IconKeys } from '../Icon/Icon';

type ButtonVariant =
  | 'default'
  | 'negative'
  | 'secondary'
  | 'secondary-noborder'
  | 'tertiary'
  | 'unstyled';

export interface ButtonProps extends CompButtonProps, ThemeComponent {
  /**
   * Set width to grow all available space
   */
  fullWidth?: boolean;
  /**
   * 'default' | 'negative' | 'secondary' | 'secondary-noborder' | 'tertiary'
   * @default default
   */
  variant?: ButtonVariant;
  /**
   * Icon from suomifi-theme
   */
  icon?: IconKeys;
  /**
   * Icon from suomifi-theme to be placed on right side
   */
  iconRight?: IconKeys;
  /**
   * Properties given to Icon-component
   */
  iconProps?: IconProps;
}

const baseClassName = 'fi-button';

const StyledButton = styled(
  ({
    theme,
    fullWidth,
    variant,
    className,
    ...passProps
  }: ButtonProps & { right?: boolean }) => (
    <CompButton
      {...passProps}
      className={classnames(className, baseClassName)}
    />
  ),
)`
  label: ${({ disabled }) =>
    disabled ? `${baseClassName}--disabled` : baseClassName};
  ${props => baseStyles(props)}
`;

const StyledIcon = styled(
  ({ right, ...passProps }: IconProps & { right?: boolean }) => (
    <Icon {...passProps} />
  ),
)`
  ${props => iconBaseStyles(props)}
`;

const iconColor = ({
  theme,
  invert,
  disabled,
}: {
  theme?: ThemeProp;
  invert?: boolean;
  disabled?: boolean;
}) => {
  if (!!theme) {
    const defaultColor = !!disabled
      ? theme.colors.disabledColor
      : theme.colors.white;
    const secondaryColor = !!disabled
      ? theme.colors.disabledColor
      : theme.colors.secondaryColor;
    return !!invert ? secondaryColor : defaultColor;
  }
  return undefined;
};

const UnstyledButton = (props: ButtonProps) => {
  const {
    className,
    theme: dissmissTheme,
    fullWidth: dissmissFullWidth,
    variant: dissmissVariant,
    icon: dismissIcon,
    iconRight: dissmissIconRight,
    ...passProps
  } = props;
  return (
    <CompButton
      {...passProps}
      className={classnames(className, baseClassName)}
    />
  );
};

class ButtonWithIcon extends Component<ButtonProps> {
  render() {
    const {
      children,
      icon,
      iconRight,
      iconProps,
      ...passProps
    } = withDefaultTheme(this.props);
    const { theme, disabled, variant } = passProps;

    if (variant === 'unstyled') {
      return <UnstyledButton {...passProps} />;
    }

    const secondaryOrTertiary =
      !!variant &&
      (variant === 'secondary' ||
        variant === 'secondary-noborder' ||
        variant === 'tertiary');
    return (
      <StyledButton {...passProps}>
        {!!icon && (
          <StyledIcon
            icon={icon}
            right={false}
            color={iconColor({ theme, disabled, invert: secondaryOrTertiary })}
            {...iconProps}
          />
        )}
        {children}
        {!!iconRight && (
          <StyledIcon
            icon={iconRight}
            right={true}
            color={iconColor({ theme, disabled, invert: secondaryOrTertiary })}
            {...iconProps}
          />
        )}
      </StyledButton>
    );
  }
}

/**
 * Use for inside Application onClick events.<br />
 * When using Button.secondaryNoborder with other than white background,<br />
 * define styles background color for all needed states (:hover, :active, :disabled)
 */
export class Button extends Component<ButtonProps> {
  static negative = (props: ButtonProps) => {
    return <ButtonWithIcon {...props} variant="negative" />;
  };

  static secondary = (props: ButtonProps) => {
    return <ButtonWithIcon {...props} variant="secondary" />;
  };

  static secondaryNoborder = (props: ButtonProps) => {
    return <ButtonWithIcon {...props} variant="secondary-noborder" />;
  };

  static tertiary = (props: ButtonProps) => {
    return <ButtonWithIcon {...props} variant="tertiary" />;
  };

  static unstyled = (props: ButtonProps) => {
    return <UnstyledButton {...props} />;
  };

  render() {
    return <ButtonWithIcon {...this.props} />;
  }
}
