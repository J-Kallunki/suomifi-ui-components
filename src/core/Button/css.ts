import { cssFromBaseStyles } from '../utils';
import { baseStyles } from './Button.baseStyles';

export const css = cssFromBaseStyles(baseStyles);
