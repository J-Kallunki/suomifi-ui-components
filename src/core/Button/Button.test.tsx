import React from 'react';
import server from 'react-dom/server';
import { render } from 'react-testing-library';
import { axe } from 'jest-axe';

import { Button } from './Button';
import { cssFromBaseStyles } from '../utils';
import { baseStyles } from './Button.baseStyles';

test('calling render with the same component on the same container does not remount', () => {
  const buttonRendered = render(<Button data-testid="button">Test</Button>);
  const { getByTestId, container, rerender } = buttonRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('button').textContent).toBe('Test');

  // re-render the same component with different props
  rerender(<Button data-testid="nottub">Test two</Button>);
  expect(getByTestId('nottub').textContent).toBe('Test two');
});

test('should not have basic accessibility issues', async () => {
  const html = server.renderToString(
    <Button data-testid="button">Test</Button>,
  );
  const results = await axe(html);
  expect(results).toHaveNoViolations();
});

test('CSS export', () => {
  const css = cssFromBaseStyles(baseStyles);
  expect(css).toEqual(expect.stringContaining('color'));
});
