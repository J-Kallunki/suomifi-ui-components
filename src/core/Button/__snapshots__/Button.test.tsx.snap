// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
  color: #282828;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 16px;
  font-weight: 600;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  padding: 8px 16px;
  min-height: 40px;
  color: #FFFFFF;
  background: linear-gradient(0deg,#2A6EBB 0%,#4387d5 100%);
  border-radius: 2px;
  text-align: center;
  text-shadow: 0 1px 1px #00357A;
  cursor: pointer;
}

.emotion-0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.emotion-0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.emotion-0::-webkit-inner-spin-button {
  height: auto;
}

.emotion-0::-webkit-outer-spin-button {
  height: auto;
}

.emotion-0:focus {
  outline: 0;
  position: relative;
}

.emotion-0:focus:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid #E97025;
  box-sizing: border-box;
  box-shadow: 0 0 10px 0 #E97025;
  z-index: 9999;
}

.emotion-0:focus:not(:focus-visible):after {
  content: none;
}

.emotion-0:hover {
  background: linear-gradient(0deg,#2E78CC 0%,#5492d9 100%);
}

.emotion-0:active {
  background: linear-gradient(0deg,#235A9A 0%,#2c72c4 100%);
}

.emotion-0.fi-button--disabled,
.emotion-0[disabled],
.emotion-0:disabled {
  text-shadow: 0 1px 1px #282828;
  background: linear-gradient(0deg,#A5ACB1 0%,#c0c5c9 100%);
  pointer-events: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0.fi-button--negative {
  background: none;
  background-color: #2A6EBB;
  border: 1px solid #FFFFFF;
  text-shadow: none;
}

.emotion-0.fi-button--negative:hover {
  background: linear-gradient(-180deg,rgba(255,255,255,0.1) 0%,rgba(255,255,255,0) 100%);
}

.emotion-0.fi-button--negative:active {
  background: none;
  background-color: #2A6EBB;
}

.emotion-0.fi-button--negative[disabled],
.emotion-0.fi-button--negative:disabled {
  opacity: 0.5;
  background: none;
}

.emotion-0.fi-button--secondary {
  color: #2A6EBB;
  background: none;
  background-color: #FFFFFF;
  border: 1px solid #2A6EBB;
  text-shadow: none;
}

.emotion-0.fi-button--secondary:hover {
  background: linear-gradient(0deg,#ECEDEE 0%,#FFFFFF 100%);
}

.emotion-0.fi-button--secondary:active {
  background: none;
  background-color: #F7FAFD;
}

.emotion-0.fi-button--secondary[disabled],
.emotion-0.fi-button--secondary:disabled {
  color: #A5ACB1;
  text-shadow: none;
  background: none;
  background-color: #F7FAFD;
  border-color: #A5ACB1;
}

.emotion-0.fi-button--secondary-noborder {
  color: #2A6EBB;
  background: none;
  background-color: #FFFFFF;
  border: 1px solid #2A6EBB;
  text-shadow: none;
  border: none;
}

.emotion-0.fi-button--secondary-noborder:hover {
  background: linear-gradient(0deg,#ECEDEE 0%,#FFFFFF 100%);
}

.emotion-0.fi-button--secondary-noborder:active {
  background: none;
  background-color: #F7FAFD;
}

.emotion-0.fi-button--secondary-noborder[disabled],
.emotion-0.fi-button--secondary-noborder:disabled {
  color: #A5ACB1;
  text-shadow: none;
  background: none;
  background-color: #F7FAFD;
  border-color: #A5ACB1;
}

.emotion-0.fi-button--tertiary {
  color: #2A6EBB;
  background: none;
  background-color: #FFFFFF;
  border: 1px solid #2A6EBB;
  text-shadow: none;
  background: #EAF2FA;
  border: none;
}

.emotion-0.fi-button--tertiary:hover {
  background: linear-gradient(0deg,#ECEDEE 0%,#FFFFFF 100%);
}

.emotion-0.fi-button--tertiary:active {
  background: none;
  background-color: #F7FAFD;
}

.emotion-0.fi-button--tertiary[disabled],
.emotion-0.fi-button--tertiary:disabled {
  color: #A5ACB1;
  text-shadow: none;
  background: none;
  background-color: #F7FAFD;
  border-color: #A5ACB1;
}

.emotion-0.fi-button--tertiary:hover {
  background: #F7FAFD;
}

.emotion-0.fi-button--tertiary:active {
  background: #F7FAFD;
}

.emotion-0 > .fi-button-icon {
  width: 16px;
  height: 16px;
  margin-right: 8px;
}

.emotion-0 > .fi-button-icon.fi-button-icon--right {
  margin-right: 0;
  margin-left: 8px;
}

<button
  aria-disabled="false"
  class="fi-button emotion-0"
  data-testid="button"
  tabindex="0"
  type="button"
>
  Test
</button>
`;
